pkg/products/grafana/reconciler.go

```diff
@@ -267,34 +268,72 @@ func (r *Reconciler) reconcileGrafanaDashboards(ctx context.Context, serverClien
+	csvName := "grafana-operator.v3.10.4"
 
-	r.log.Info("custom uninstallation of 3.10.4 of grafana operator to allow upgrade to 4.x.x")
-	subscription := &v1alpha1.Subscription{
+	// Confirm that the upgrade is from 3.10.4 to 4.2.0 using the Grafana Operator CSV
+	csv := &v1alpha1.ClusterServiceVersion{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      constants.GrafanaSubscriptionName,
-			Namespace: operatorNamespace,
+			Name:      csvName,
+			Namespace: namespace,
 		}}
 
-	err := client.Get(ctx, k8sclient.ObjectKey{Name: constants.GrafanaSubscriptionName, Namespace: operatorNamespace}, subscription)
+	err := client.Get(ctx, k8sclient.ObjectKey{Name: csv.Name, Namespace: csv.Namespace}, csv)
 	if err != nil && !k8serr.IsNotFound(err) {
 		return integreatlyv1alpha1.PhaseFailed, err
 	}
-	// if the rhmi-grafana subscription is not found in the cluster, we are in install so we don't need to try to delete the older version
-	// return phasecompleted
-	if k8serr.IsNotFound(err) {
-		return integreatlyv1alpha1.PhaseCompleted, nil
-	}
 
-	// if the rhmi-grafana subscription is not 3.10.4 we'll return, it means we have moved to 4.2.0
-	if subscription.Status.InstalledCSV != "grafana-operator.v3.10.4" {
-		return integreatlyv1alpha1.PhaseCompleted, nil
+	// If the grafana-operator.v3.10.4 CSV exists and is in the process of being replaced, scale the deployment to 0
+	if csv.Status.Phase == "Replacing" {
+		phase, err := r.scaleDeployment(ctx, client, deploymentName, namespace, 0)
+		if err != nil {
+			return phase, nil
+		}
 	}
 
-	err = client.Delete(ctx, subscription)
+	// Check if the deployment is scaled to 0, and if so, scale it back up to 1
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deploymentName,
+			Namespace: namespace,
+		},
+	}
+	err = client.Get(ctx, k8sclient.ObjectKey{Name: deployment.Name, Namespace: deployment.Namespace}, deployment)
 	if err != nil && !k8serr.IsNotFound(err) {
 		return integreatlyv1alpha1.PhaseFailed, err
 	}
+	if deployment.Status.Replicas == 0 {
+		phase, err := r.scaleDeployment(ctx, client, deploymentName, namespace, 1)
+		if err != nil {
+			return phase, nil
+		}
+	}
+
+	return integreatlyv1alpha1.PhaseCompleted, nil
+}
+
+func (r *Reconciler) scaleDeployment(ctx context.Context, client k8sclient.Client, name string, namespace string, scaleValue int32) (integreatlyv1alpha1.StatusPhase, error) {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		deployment := &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		}
+		err := client.Get(ctx, k8sclient.ObjectKey{Name: deployment.Name, Namespace: deployment.Namespace}, deployment)
+		if err != nil {
+			return fmt.Errorf("failed to get DeploymentConfig %s in namespace %s with error: %s", deployment.Name, namespace, err)
+		}
+
+		deployment.Spec.Replicas = &scaleValue
+		err = client.Update(ctx, deployment)
+		return err
```

Comment:
only return here if if err != nil

Reply:
@laurafitzgerald I could be wrong but I think this is the correct behavior because the err being returned here is only for the `RetryOnConflict` update function. If err == nil then it breaks out of the `RetryOnConflict` block but if err != nil then it backs off and tries again until it retries too many times. 

That being said, if we want to update it for clarity/readability I can definitely change it to something like:
```
if err != nil {
    return err
}
return nil
```

Reply:
Thanks Carl. No need for the update. I see that it's the end of the function. 👍 
Would it have been an option to use CreateOrUpdate function here? I've never seen retryoncnflict before. 🤔

Reply:
I forget if I tried using CreateOrUpdate at one point but I ultimately went with RetryOnConflict because it was used elsewhere in the codebase when working with Deployments. I believe one of the benefits of RetryOnConflict is that it uses exponential backoff when trying to get the latest version of the deployment which helps prevent exhausting the apiserver. Not sure if CreateOrUpdate also uses this or if we need to worry about that for one call but it's worth mentioning.